src/allocator/mod.rs

```diff
@@ -0,0 +1,146 @@
+//! Implementation of the HermitCore Allocator for dynamically allocating heap memory
+//! in the kernel.
+
+mod bootstrap;
+mod bump;
+
+use core::alloc::{AllocError, Allocator, GlobalAlloc, Layout};
+use core::ptr;
+use core::ptr::NonNull;
+
+use align_address::Align;
+use hermit_sync::InterruptTicketMutex;
+use log::info;
+use talc::{ErrOnOom, Span, Talc};
+
+use self::bootstrap::BootstrapAllocator;
+use self::bump::BumpAllocator;
+
+/// The global system allocator for Hermit.
+struct GlobalAllocator {
+	/// The bootstrap allocator, which is available immediately.
+	///
+	/// It allows allocations before the heap has been initalized.
+	bootstrap_allocator: Option<BootstrapAllocator<BumpAllocator>>,
+
+	/// The heap allocator.
+	///
+	/// This is not available immediately and must be initialized ([`Self::init`]).
+	heap: Option<Talc<ErrOnOom>>,
+}
+
+impl GlobalAllocator {
+	const fn empty() -> Self {
+		Self {
+			bootstrap_allocator: None,
+			heap: None,
+		}
+	}
+
+	/// Initializes the heap allocator.
+	///
+	/// # Safety
+	///
+	/// The memory starting from `heap_bottom` with a size of `heap_size`
+	/// must be valid and ready to be managed and allocated from.
+	unsafe fn init(&mut self, heap_bottom: *mut u8, heap_size: usize) {
+		let arena = Span::from_base_size(heap_bottom, heap_size);
+		let mut talc = Talc::new(ErrOnOom);
+
+		unsafe { talc.claim(arena); }
+
+		self.heap = Some(talc);
+	}
+
+	fn align_layout(layout: Layout) -> Layout {
+		let size = layout
+			.size()
+			.align_up(core::mem::size_of::<crossbeam_utils::CachePadded<u8>>());
+		let align = layout
+			.align()
+			.max(core::mem::align_of::<crossbeam_utils::CachePadded<u8>>());
+		Layout::from_size_align(size, align).unwrap()
+	}
```

Comment:
We can skip overaligning here, since there will never be contention on the allocations. This way we can also avoid the crossbeam-utils dependency.